### PR TITLE
Refactor `LoginViewModel` to create a new sitter or owner

### LINKED ISSUE #172: Refactor login process to create and store the new instance of sitter or owner based on the selected role.
## Description
After the user selects their role on the login screen (tapped on the button `I am an owner` or `I am a sitter`), the new instance of owner or sitter should be created and saved to persistent storage based on the selected role.

### Dev requirements and restrictions:
When the user selects `owner` role, the new instance of the owner should be created and saved to persistent storage `KeyValueStorage.Name.currentOwner` with the key `KeyValueStorage.Key.currentOwner`.

When the user selects `sitter` role, the new instance of the sitter should be created and saved to persistent storage `KeyValueStorage.Name.currentSitter` with the key `KeyValueStorage.Key.currentSitter`.

### PR DESCRIPTION
**Pull request by marinkman**
## Summary
This PR is a part of the work to resolve #172 

## Changes
- added published properties  to the `LoginViewModel` to determine which role the user has chosen
- applied the published properties of the `LoginViewModel` in the `LoginView`
- added private methods to the `LoginViewModel` to create a new owner or sitter

## Checks to complete
- [x] I've built and run my changes, and no warnings or errors occurred.
- [x] All existing tests passed with my changes.
- [x] My code follows our style guide.
- [x] I've performed a self-review of my code.
- [x] PR's title has a conventional commit style (short and descriptive).
- [x] Every non-private interface is documented properly.
- [x] I've added tests for my code (if it makes sense).
- [x] PR has assignee, reviewers, milestone, it is properly linked to the project and to the issue.

### CODE REVIEW

File: Woof/Woof/ViewModel/LoginViewModel.swift [hunk 1]
```diff
@@ -1,11 +1,76 @@
...
+    }
+
+    // MARK: - Private interface
+
+    private var isTheCurrentOwnerExists: Bool {
+        guard let data = KeyValueStorage(KeyValueStorage.Name.currentOwner)
+            .loadData(for: KeyValueStorage.Key.currentOwner) else {
+            return false
+        }
+        guard let owner = try? JSONDecoder().decode(Owner.self, from: data) else {
```
Comment (YuliaNastiushkina) [inline Woof/Woof/ViewModel/LoginViewModel.swift:36]: I see a violation on this line, check it, please.

File: Woof/Woof/ViewModel/LoginViewModel.swift [hunk 2]
```diff
@@ -1,11 +1,76 @@
...
+
+        return true
+    }
+
+    private var isTheCurrentSitterExists: Bool {
+        guard let data = KeyValueStorage(KeyValueStorage.Name.currentSitter)
+            .loadData(for: KeyValueStorage.Key.currentSitter) else {
+            return false
+        }
+        guard let sitter = try? JSONDecoder().decode(Sitter.self, from: data) else {
```
Comment (YuliaNastiushkina) [inline Woof/Woof/ViewModel/LoginViewModel.swift:48]: I see a violation on this line, check it, please.

**Review by YuliaNastiushkina — CHANGES_REQUESTED** [review]
Very good!
Check a few nits and add information to the CHANGELOG, please.

File: Woof/Woof/ViewModel/LoginViewModel.swift (hunk 1 above)
Comment (marinkman) [reply Woof/Woof/ViewModel/LoginViewModel.swift]: Thanks, fixed.

File: Woof/Woof/ViewModel/LoginViewModel.swift (hunk 2 above)
Comment (marinkman) [reply Woof/Woof/ViewModel/LoginViewModel.swift]: Thanks, fixed.

**Review by iBYarchitect — COMMENTED** [review]
Is it worth testing the behavior of the isTheOwnerRoleSelected and isTheSitterRoleSelected properties when they are set to true?

File: Woof/WoofTests/ViewModel/LoginViewModelTests.swift [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    }
+
+    func testLoginViewModelIsInitializedWithExpectedValues() {
+        // Given // When
+        let viewModel = LoginViewModel()
+
+        // Then
+        XCTAssertEqual(viewModel.isTheOwnerRoleSelected, false)
+        XCTAssertEqual(viewModel.isTheSitterRoleSelected, false)
+    }
```
Comment (NadzeyaShpakouskaya) [inline Woof/WoofTests/ViewModel/LoginViewModelTests.swift:15]: Can we simplify it?
```suggestion
        XCTAssertFalse(viewModel.isTheOwnerRoleSelected)
        XCTAssertFalse(viewModel.isTheSitterRoleSelected)
 
```

File: Woof/WoofTests/ViewModel/LoginViewModelTests.swift [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import XCTest
+
+final class LoginViewModelTests: XCTestCase {
```
Comment (NadzeyaShpakouskaya) [inline Woof/WoofTests/ViewModel/LoginViewModelTests.swift:3]: Add more tests for properties:

- @Published var isTheOwnerRoleSelected
- @Published var isTheSitterRoleSelected

**Review by NadzeyaShpakouskaya — CHANGES_REQUESTED** [review]
Well done 🥳 
Some minor improvements:

File: Woof/Woof/ViewModel/LoginViewModel.swift [hunk 5]
```diff
@@ -1,11 +1,76 @@
...
+    ///
+    /// Returns `true` if the owner role is selected, otherwise `false`.
+    ///
+    /// The property publishes updates when the value is changed.
+    @Published var isTheOwnerRoleSelected = false {
+        didSet {
+            createNewOwner()
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (marinkman) [reply Woof/Woof/ViewModel/LoginViewModel.swift]: I already fixed it.

**Review by YuliaNastiushkina — APPROVED** [review]
Overall LGTM.
Just answer other reviewers and their comments.

File: Woof/WoofTests/ViewModel/LoginViewModelTests.swift (hunk 3 above)
Comment (marinkman) [reply Woof/WoofTests/ViewModel/LoginViewModelTests.swift]: Done.

File: Woof/WoofTests/ViewModel/LoginViewModelTests.swift (hunk 4 above)
Comment (marinkman) [reply Woof/WoofTests/ViewModel/LoginViewModelTests.swift:3]: Done.

File: Woof/Woof/ViewModel/LoginViewModel.swift [hunk 6]
```diff
@@ -1,11 +1,76 @@
...
+        didSet {
+            if !isTheCurrentSitterExists {
+                createAndSaveNewSitter()
+            }
+        }
+    }
+
+    // MARK: - Private interface
+
+    private var isTheCurrentOwnerExists: Bool {
```
Comment (NadzeyaShpakouskaya) [inline Woof/Woof/ViewModel/LoginViewModel.swift:35]: nit:
```suggestion

    private var isCurrentOwnerExists: Bool {
```

File: Woof/Woof/ViewModel/LoginViewModel.swift [hunk 7]
```diff
@@ -1,11 +1,76 @@
...
+            return false
+        }
+        guard (try? JSONDecoder().decode(Sitter.self, from: data)) != nil else {
+            return false
+        }
+
+        return true
+    }
+
+    private func createAndSaveNewOwner() {
```
Comment (NadzeyaShpakouskaya) [inline Woof/Woof/ViewModel/LoginViewModel.swift:59]: nit:
using the word `and` in method naming is unusual. It is strange if we can save objects without creating.
```suggestion

    private func saveNewOwner() {
```

**Review by NadzeyaShpakouskaya — APPROVED** [review]
Overall LGTM

File: Woof/Woof/ViewModel/LoginViewModel.swift (hunk 7 above)
Comment (marinkman) [reply Woof/Woof/ViewModel/LoginViewModel.swift:59]: Fixed.

File: Woof/WoofTests/ViewModel/LoginViewModelTests.swift [hunk 8]
```diff
@@ -1,16 +1,94 @@
...
     }
 
     func testLoginViewModelIsInitializedWithExpectedValues() {
-        // Given // When
-        let viewModel = LoginViewModel()
+        XCTAssertFalse(viewModel.isTheOwnerRoleSelected)
+        XCTAssertFalse(viewModel.isTheSitterRoleSelected)
+    }
+
+    func testWhenIsTheOwnerRoleSelectedIsTrueTheNewOwnerIsCreatedIfTheCurrentOwnerDoesNotExist() {
```
Comment (iBYarchitect) [inline Woof/WoofTests/ViewModel/LoginViewModelTests.swift:25]:
```suggestion
    func testWhenIsTheOwnerRoleSelectedToTrueTheNewOwnerIsCreatedIfTheCurrentOwnerDoesNotExist() {
```

File: Woof/WoofTests/ViewModel/LoginViewModelTests.swift [hunk 9]
```diff
@@ -1,16 +1,94 @@
...
+
+        let newOwner = KeyValueStorage(KeyValueStorage.Name.currentOwner)
+            .loadData(for: KeyValueStorage.Key.currentOwner)
+
+        // Then
+        XCTAssertNil(currentOwner)
+        XCTAssertNotNil(newOwner)
+    }
+
+    func testWhenIsTheOwnerRoleSelectedIsTrueNoNewOwnerIsCreatedIfTheCurrentOwnerAlreadyExist() {
```
Comment (iBYarchitect) [inline Woof/WoofTests/ViewModel/LoginViewModelTests.swift:41]:
```suggestion
    func testWhenIsTheOwnerRoleSelectedToTrueNoNewOwnerIsCreatedIfTheCurrentOwnerAlreadyExist() {
```

**Review by iBYarchitect — CHANGES_REQUESTED** [review]
All is well, there is a small note on the name of the tests:

File: Woof/WoofTests/ViewModel/LoginViewModelTests.swift (hunk 8 above)
Comment (marinkman) [reply Woof/WoofTests/ViewModel/LoginViewModelTests.swift]: Corrected test method name.

File: Woof/WoofTests/ViewModel/LoginViewModelTests.swift (hunk 9 above)
Comment (marinkman) [reply Woof/WoofTests/ViewModel/LoginViewModelTests.swift]: Corrected test method name.

**Review by iBYarchitect — APPROVED** [review]
LGTM

**Pull request merged**